lib/kubernetes-deploy/kubernetes_resource/replica_set.rb

```diff
@@ -53,6 +53,17 @@ def available_replicas
 
     private
 
+    def stale_status?
+      observed_generation != current_generation
+    end
+
+    def check_pods?
+      # We're only using pods in deploy_failed? to check that they aren't ALL bad,
+      # so if we can already tell that from the RS data, don't bother looking at (or expensively fetching) them
+      return false if !exists? || stale_status?
+      ready_replicas < [2, desired_replicas].min
```

Comment:
> because we only need to find a single one that didn't fail in order to know that not all of them failed

Technically true, but workloads that do not have proper health checks (e.g. a lot of apps' jobs deployments) will very briefly appear to come up as healthy even if they crash very soon afterwards. Requiring us to see two healthy rather than one before making this assumption is intended as a small safeguard against that scenario. Since the bypass this PR adds is most useful at the tail end of large rollouts, we could choose an even higher number and still get the benefit when it matters most. But the sooner we stop needing pods for _any_ deployment in the set, the more effective it will be overall.

> Why do we care about desired_replicas here?

Because deployments can have less than 2 replicas. If that happens, `ready_replicas < 2` will always be true even when the deployment is completely rolled out, and we'll be fetching the pods unconditionally. We definitely don't want to do that for zero-replica deployments, but maybe this would be more sensible:

```ruby
return false unless exists?
return false if stale_status?
return false if desired_replicas == 0
ready_replicas > 1
```

i.e. if the deployment only has one replica, that replica is pretty important and worth checking unconditionally.

> Also, is a test going to fail if I change this to just ready_replicas < 2

The `test_sync_does_not_request_pods_if_desired_replicas_is_zero` was supposed to fail if you ignore `desired_replicas`, but I forgot to set the generation in the status, so it wasn't. Will push a fix.

>  (or even < 1)?

Both new tests fail if you do this (and one still fails if you just change the 2 to a 1 without ignoring desired replicas).

Reply:
>  will very briefly appear to come up as healthy even if they crash very soon afterwards. Requiring us to see two healthy rather than one before making this assumption is intended as a small safeguard against that scenario.

My concern with this PR was `2` seemed like a magic number. Given the above did you consider also making this time based? We've done something like that in `Job#running?`

Reply:
> Given the above did you consider also making this time based? 

I don't think we have a relevant timestamp to use in this case. ReplicaSet supports status conditions, but it doesn't have any that are always (or even usually) present. The only timestamp I'm seeing is the creation time, which isn't useful because RS can be reused (e.g. during rollbacks).